### PR TITLE
Feature/notifications branch

### PR DESCRIPTION
**Pull request by UgurDenli**
26e0a4f0fbc514c36672e498c6c6875b

### CODE REVIEW

File: WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt [hunk 1]
```diff
@@ -1,19 +1,250 @@
...
+import android.view.LayoutInflater
+import android.view.View
+import androidx.core.content.ContextCompat.getSystemService
 import androidx.navigation.NavController
 import com.example.whatsapp.R
+import com.example.whatsapp.base.BaseFragment
+import com.example.whatsapp.databinding.FragmentNotificationsBinding
 import dagger.hilt.android.AndroidEntryPoint
 
+
```
Comment (OskarLasota) [inline WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt:20]: ctrl + alt + o and ctrl + alt + l to remove unused imports and reformat

File: WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt [hunk 2]
```diff
@@ -1,19 +1,250 @@
...
+            notificationChannel.lightColor = Color.RED
+            notificationChannel.vibrationPattern = longArrayOf(0, 1000, 500, 1000)
+            notificationChannel.enableVibration(true)
+            notificationManager.createNotificationChannel(notificationChannel)
+        }
+
+    }
+    private fun observeUiInteractions() {
+        //Message Layouts
+        binding.vibrateLayout.setOnClickListener {
```
Comment (OskarLasota) [inline WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt:78]: this is fine, although i'd reduce repeated code and also the size of thie entire function. I'd do something like : 

binding.vibrateLayout.setonclicklistener { 
     showOptionsDialog()
}

private fun showOptionsDialog(title : String, choices : List<String or @ResStringId>, ) { 
```suggestion
        binding.vibrateLayout.setOnClickListener {
```
            val builder = AlertDialog.Builder(context)
            builder.setTitle("Vibrate")
            val vibrates = arrayOf("Off", "Default", "Short", "Long")
            val checkedItem = 1
            builder.setSingleChoiceItems(vibrates, checkedItem) { dialog, which ->
                val position = (dialog as AlertDialog).listView.checkedItemPosition
                when(position){
                0 -> //
                1 -> //
                }

                dialog.dismiss()
            }
            val dialog = builder.create()
            dialog.show()
} 

basically make this function reusable

File: WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt [hunk 3]
```diff
@@ -1,19 +1,250 @@
...
+                    binding.vibrateDesc.text = getString(R.string.tvlong)
+                }
+                dialog.dismiss()
+            }
+            val dialog = builder.create()
+            dialog.show()
+        }
+        binding.popUpLayout.setOnClickListener {
+            val builder = AlertDialog.Builder(context)
+            builder.setTitle("Popup notification")
```
Comment (OskarLasota) [inline WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt]: avoid hardcoded strings

File: WhatsApp/app/src/main/res/layout/fragment_notifications.xml [hunk 4]
```diff
@@ -1,6 +1,564 @@
 <?xml version="1.0" encoding="utf-8"?>
-<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+
+<ScrollView
```
Comment (OskarLasota) [inline WhatsApp/app/src/main/res/layout/fragment_notifications.xml]: ctrl + alt + l

File: WhatsApp/app/src/main/res/layout/fragment_notifications.xml [hunk 5]
```diff
@@ -1,6 +1,564 @@
...
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:id="@+id/root_layout"
```
Comment (OskarLasota) [inline WhatsApp/app/src/main/res/layout/fragment_notifications.xml]: dont forget to follow the naming convention. Should be svNotifications or something alike

File: WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt [hunk 6]
```diff
@@ -1,19 +1,250 @@
...
+            val popup = arrayOf("No popup", "Only when screen is 'on' ", "Only when screen is 'off'", "Always show popup")
+            val checkedItem = 1
```
Comment (OskarLasota) [inline WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt]: should be 0 right? if thats the default

File: WhatsApp/app/src/main/res/layout/fragment_notifications.xml [hunk 7]
```diff
@@ -1,6 +1,564 @@
...
+                android:layout_marginBottom="4dp"
+                android:text="@string/popupnotification" />
+
+            <TextView
+                android:id="@+id/tvPopUpTonesDesc"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_below="@+id/tvPopUpTones"
+                android:layout_marginStart="16dp"
+                android:text="@string/nopopup" />
```
Comment (OskarLasota) [inline WhatsApp/app/src/main/res/layout/fragment_notifications.xml]: I believe these text views should be initialized in the fragment and the text should be obtained through some sort of local cache like shared preferences or data store. We don't want to hard code the default here.

File: WhatsApp/app/src/main/res/layout/fragment_notifications.xml [hunk 8]
```diff
@@ -1,6 +1,564 @@
...
+
+        </RelativeLayout>
+
+        <RelativeLayout
+            android:id="@+id/lightLayout"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:clickable="true"
+            android:focusable="true"
```
Comment (OskarLasota) [inline WhatsApp/app/src/main/res/layout/fragment_notifications.xml:173]: not sure, is focusable required in these layouts? dont remember what it does

File: WhatsApp/app/src/main/res/layout/fragment_notifications.xml [hunk 9]
```diff
@@ -1,6 +1,564 @@
...
-<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+
+<ScrollView
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:id="@+id/root_layout"
     android:layout_width="match_parent"
-    android:layout_height="match_parent">
+    android:layout_height="wrap_content">
+
```
Comment (OskarLasota) [inline WhatsApp/app/src/main/res/layout/fragment_notifications.xml:9]: image

you can see on the right the warning signs. Fix those please

File: WhatsApp/app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -32,4 +57,22 @@
     <string name="changenumber">Change number</string>
     <string name="requestaccountinfo">Request account info</string>
     <string name="deletemyaccount">Delete my account</string>
+
+    <!-- Strings used for notification page in settings area -->
+    <string name="conversationtones">Conversation tones</string>
+    <string name="playsounds">Play sounds for incoming and outgoing\nmessages</string>
```
Comment (OskarLasota) [inline WhatsApp/app/src/main/res/values/strings.xml]: Whatsapp did not use the newline /n approach. They use margin i believe. Would look weird on a horizontal orientation.

File: WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt (hunk 2 above)
Comment (Josh-Owen) [reply WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt:78]: We should avoid using hard coded text, also this can be simplified to something like below:

```     
        val vibrates = arrayOf("Off", "Default", "Short", "Long")
        val checkedItem = 1
        AlertDialog.Builder(requireContext())
            .setTitle("Vibrate")
            .setSingleChoiceItems(vibrates, checkedItem) { dialog, position ->
                when (position) {
                    0 -> {}
                    1 -> {}
                }
                dialog.dismiss()
            }
            .create()
            .show()
```
            
Note: Also look at how you were trying to find the position of the selected item in the list view in addition to the other changes. You should try to make these changes to the other alert dialogs also. 

Also, I think it's a good shout if instead of having all of the application defaults settings in random functions in the app we should store them in some configuration file which is solely responsible for housing all of our application default values.

File: WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt (hunk 1 above)
Comment (UgurDenli) [reply WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt:20]: done :)

File: WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt (hunk 6 above)
Comment (UgurDenli) [reply WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt]: number doesnt matter too much it will be changed when the implementation of the feature gets added

File: WhatsApp/app/src/main/res/layout/fragment_notifications.xml (hunk 7 above)
Comment (UgurDenli) [reply WhatsApp/app/src/main/res/layout/fragment_notifications.xml]: its just here until we add the features and im going to take it out

File: WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt [hunk 11]
```diff
@@ -1,19 +1,144 @@
...
+    }
 
 
+    //    /**
+//     * Vibrates the device. Used for providing feedback when the user performs an action.
+//     */
+//    fun vibrate(view: View) {
+//        view.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
+//    }
+//    private fun defaultLed(){
```
Comment (NelzkieLabs) [inline WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/settings/notifications/NotificationsFragment.kt]: This commented code, is it still needed for a reference?

**Review by OskarLasota — APPROVED** [review]
